src/network/networkd-dhcp4.c

```diff
@@ -155,11 +160,21 @@ static int link_set_dhcp_routes(Link *link) {
                 if (IN_SET(route->scope, RT_SCOPE_LINK, RT_SCOPE_UNIVERSE))
                         route->prefsrc.in = address;
 
+                if (set_contains(link->dhcp_routes, route))
+                        continue;
+
                 r = route_configure(route, link, dhcp4_route_handler);
                 if (r < 0)
                         return log_link_error_errno(link, r, "Could not set host route: %m");
-                if (r > 0)
+                if (r > 0) {
                         link->dhcp4_messages++;
+
+                        r = set_put(link->dhcp_routes, route);
+                        if (r < 0)
+                                return log_link_error_errno(link, r, "Failed to store DHCP route: %m");
+
+                        TAKE_PTR(route);
+                }
```

Comment:
This pattern with route_configure and set_put appears three times in the patch. Maybe define a helper func for it?